distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/KadAction.java

```diff
@@ -0,0 +1,369 @@
+         */
+        public int getCode() {
+            return code;
+        }
+
+        /**
+         * @param code the code for the Action
+         * @return the ActionType with the corresponding code, or invalid if an invalid code is passed.
+         */
+        public static PayloadType getTypeFromCode(int code) {
+            for (PayloadType type : PayloadType.values()) {
+                if (type.code == code) return type;
+            }
+            return INVALID;
+        }
+    }
+
+    /**
+     *  Constructor of KadAction.
+     *
+     * @param actionType The type of the action.
+     * @param id The id of the message.
+     * @param part The number of the current message part.
+     * @param maxParts The number of the total messages.
+     * @param payloadType The type of the payload.
+     * @param payload The value of the payload.
+     * @throws IllegalArgumentException if the parameters are not valid.
+     */
+    public KadAction(@NonNull SMSPeer smsPeer, @NonNull ActionType actionType, int id, int part, int maxParts, @NonNull PayloadType payloadType, @NonNull String payload) {
+        this.actionPeer = smsPeer;
+        this.actionType = actionType;
+        this.operationId = id;
+        this.currentPart = part;
+        this.totalParts = maxParts;
+        this.payloadType = payloadType;
+        this.payload = payload;
+        if (!isValid())
+            throw new IllegalArgumentException();
+    }
+
+    /**
+     * Constructor of KadDAction given SMSMessage.
+     *
+     * @param buildingMessage the given SMSMessage.
+     * @throws IllegalArgumentException if the parameters are not valid.
+     */
+    public KadAction(@NonNull SMSMessage buildingMessage) {
+        String messageBody = buildingMessage.getData();
+        actionPeer=buildingMessage.getPeer();
+        actionType=ActionType.getTypeFromVal(Integer.parseInt(messageBody.substring(ACTION_TYPE_START_INDEX,OPERATION_ID_START_INDEX)));
+        operationId=removePadding(messageBody.substring(OPERATION_ID_START_INDEX,CURRENT_PART_START_INDEX));
+        currentPart=removePadding(messageBody.substring(CURRENT_PART_START_INDEX,TOTAL_PARTS_START_INDEX));
+        totalParts=removePadding(messageBody.substring(TOTAL_PARTS_START_INDEX,PAYLOAD_TYPE_START_INDEX));
+        payloadType=PayloadType.getTypeFromCode(Integer.parseInt(messageBody.substring(PAYLOAD_TYPE_START_INDEX,PAYLOAD_START_INDEX)));
+        payload=messageBody.substring(PAYLOAD_START_INDEX);
+    }
+
+    /**
+     * Check if all the action's parameters are valid
+     *
+     * @return True if the defined action is valid and fits into a Message.
+     */
+    @Override
+    public boolean isValid() {
+        if (this.actionType.equals(ActionType.INVALID))
+            return false;
+        if (this.payloadType.equals(PayloadType.INVALID))
+            return false;
+        if (currentPart <MIN_PARTS || totalParts <= MIN_PARTS || currentPart > totalParts || totalParts>MAX_PARTS )
```

Comment:
Sono 35 minuti che non capisco cosa cazzo ha la demo che crasha e crasha e crasha e crasha.
totalParts <= MIN_PARTS rende invalida l'Action? Non posso inviare meno di due messaggi? Dude pliz

Reply:
it's definitely wrong, total <= 1 is not invalid if the total is actually 1, for example, ping has one message of one (1/1) and must be considered valid.
Should be totalParts < MIN_PARTS

Reply:
Dopo averlo cambiato a totalParts<Min_parts pensavo proprio di togliere quest'ultimo controllo perché se TotalParts<1 allora automaticamente si rientra nella condizione subito dopo ovvero currentPart>totalParts  essendo la condizione subito prima currentPart <MinParts non soddisfatta ;